### ISSUE TITLE
onCompleteFailure called multiple times

### ISSUE THREAD

**Issue by forfreeday** [eclipse/jetty.project#9476]
**Jetty version(s)**
Jetty 9.4.51

**Java version/vendor** `(use: java -version)`
java version 1.8.0_161

**OS type/version**
CentOS

**Description**

When I upgrade to Jetty 9.4.51, this exception will appear in the log every day.
What caused this anomaly?

```
04:04:21.744 WARN  [qtp903644315-163107] [o.e.j.i.AbstractConnection](AbstractConnection.java:96)
java.lang.NullPointerException: null
        at org.eclipse.jetty.io.AbstractConnection.lambda$failedCallback$0(AbstractConnection.java:92)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by joakime** [comment]
Jetty 9.x is now at End of Community Support.

See
* #7958

You should be upgrading to Jetty 10 or Jetty 11 at this point.

**Comment by forfreeday** [comment]
Our system is JDK8, can not upgrade to a higher version, can only use jetty9

**Comment by olamy** [comment]
> Our system is JDK8, can not upgrade to a higher version, can only use jetty9

You should definitely read https://github.com/eclipse/jetty.project/issues/7958 and how to get Jetty 9.x support.

**Comment by JoergHeinicke5005** [comment]
I do understand that you can't support multiple major versions in parallel, but you have to end support at some point. Where I'm not following, is, that a NullPointerException had recently been introduced (with v9.4.51, which means it got broken by an actively performed change, i.e. while in EOL) and nonetheless you don't want to take care of it.
If I understand correctly, this new patch version v9.4.51 is for professional support of Jetty. So, there is the slight chance that the NPE gets silently fixed the same way it got introduced. Then again, it would be easy to provide tracking of such an issue in a public ticket like this one.
Btw., if running Spring Boot 2 with Jetty, v9 is still the default version (even though it can be switched to v10). I doubt these will be the last occurrences you'll see this issue being reported.

**Comment by schrotti74** [comment]
I agree. Also recently upgraded to 9.4.51 in pre-prod (e.g. loadtest) where we have seen the NPE. If this is some sort of regression recently introduced, it should be fixed in another 9.4.x based release.

**Comment by gregw** [comment]
Since we must have introduced in recent release, we will look at fixing it as well.

**Comment by AN3Orik** [comment]
Any news?

**Comment by joakime** [comment]
@AN3Orik since Jetty 9.x is End of Community Support, this is extremely low priority.

Our overwhelming focus is on Jetty 12, with support for Jetty 10 and Jetty 11 right now.

**Comment by joakime** [comment]
This is reported against `AbstractConnection` which has had no code changes since Aug 22, 2018 

https://github.com/eclipse/jetty.project/commits/jetty-9.4.x/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractConnection.java

Which was around the release of Jetty 9.4.12.

**Comment by AN3Orik** [comment]
It's start happening after changes in this commit i think
[add some fixes to SizeLimitHandler](https://github.com/eclipse/jetty.project/commit/ec197a245e91b4d2c3b70c0809543404769bad5f)

Failed callback added and this callback is null when AbstractConnection.failedCallback() method reached

**Comment by lachlan-roberts** [comment]
@AN3Orik `SizeLimitHandler` was only added in `9.4.50` and is not added to the server by default.

You must manually add it for it to ever be used, so unless you have started using this handler in `9.4.50` it is not related.

**Comment by lachlan-roberts** [comment]
I reviewed the 9.4.50 -> 9.4.51 diff and this almost certainly comes from this change https://github.com/eclipse/jetty.project/pull/9062.

The `failedCallback()` method is only called from `onCompleteFailure`, and #9062 changes how this can be called.
@gregw @sbordet

**Comment by sid-evolphin** [comment]
> Since we must have introduced in recent release, we will look at fixing it as well.

@gregw Please review the PR #9932 I have submitted to avoid a null `Callback` from derailing the synchronous flow subsequent to the invocation of `o.e.j.i.AbstractConnection.failedCallback(Callback, Throwable)`. Also if debug logging is enabled, then the root cause may be ascertained for the NPE.

Thank you very much.

**Comment by gregw** [comment]
@sbordet @lachlan-roberts I think this is actually a real bug in IteratingCallback in all our releases.  I think we are able to call `onCompleteFailure` multiple times for the same failure.

**Comment by gregw** [comment]
@sbordet @lachlan-roberts @lorban The following "test" is passing when it should not.  It shows that we can call `onCompleteFailure` multiple times. We can also call `process` after a failure!
```java


    @Test
    public void testMultipleFailures() throws Exception
    {
        AtomicInteger process = new AtomicInteger();
        AtomicInteger failure = new AtomicInteger();
        IteratingCallback icb = new IteratingCallback()
        {
            @Override
            protected Action process() throws Throwable
            {
                process.incrementAndGet();
                return Action.SCHEDULED;
            }

            @Override
            protected void onCompleteFailure(Throwable cause)
            {
                super.onCompleteFailure(cause);
                failure.incrementAndGet();
            }
        };

        icb.iterate();
        assertEquals(1, process.get());
        assertEquals(0, failure.get());

        icb.failed(new Throwable("test1"));

        assertEquals(1, process.get());
        assertEquals(1, failure.get());

        icb.succeeded();
        assertEquals(2, process.get());
        assertEquals(1, failure.get());

        icb.failed(new Throwable("test1"));
        assertEquals(2, process.get());
        assertEquals(2, failure.get());
```

**Comment by sid-evolphin** [comment]
Thanks a lot @gregw!

**Comment by gregw** [comment]
merged through to 12.0.x

@forfreeday We are 99% sure this will fix your problem, but we did not have an exact reproducer of your exact case.  Any chance you can test this out before we do a release?

**Comment by markkolich** [comment]
@gregw are you still looking for someone to try this out pre-release?

I can reproduce this NPE on `9.4.51.v20230217` pretty much at will, so I'm happy to try a snapshot with a fix and see if that resolves the bug.

If you need a guinea pig let me know what 9.x snapshot I should try.

**Comment by gregw** [comment]
@markkolich that would be good.  I think the most recent nightly is https://oss.sonatype.org/service/local/repo_groups/jetty-with-staging/content/org/eclipse/jetty/jetty-distribution/9.4.52-SNAPSHOT/jetty-distribution-9.4.52-20230807.110254-149.tar.gz  else you can build it from source.
